src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Commands;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.ProjectModel;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace NuGet.CommandLine.XPlat
+{
+    public class AddPackageReferenceCommandRunner : IAddPackageReferenceCommandRunner
+    {
+        private const string NUGET_RESTORE_MSBUILD_VERBOSITY = "NUGET_RESTORE_MSBUILD_VERBOSITY";
+        private const int MSBUILD_WAIT_TIME = 2 * 60 * 1000; // 2 minutes in milliseconds
+
+        public async Task<int> ExecuteCommand(PackageReferenceArgs packageReferenceArgs, MSBuildAPIUtility msBuild)
+        {
+            packageReferenceArgs.Logger.LogInformation(string.Format(CultureInfo.CurrentCulture,
+                Strings.Info_AddPkgAddingReference,
+                packageReferenceArgs.PackageDependency.Id,
+                packageReferenceArgs.ProjectPath));
+
+            if (packageReferenceArgs.NoRestore)
+            {
+                packageReferenceArgs.Logger.LogWarning(string.Format(CultureInfo.CurrentCulture,
+                    Strings.Warn_AddPkgWithoutRestore));
+
+                msBuild.AddPackageReference(packageReferenceArgs.ProjectPath, packageReferenceArgs.PackageDependency);
+                return 0;
+            }
+
+            // 1. Get project dg file
+            packageReferenceArgs.Logger.LogDebug("Reading project Dependency Graph");
+            var dgSpec = ReadProjectDependencyGraph(packageReferenceArgs);
+            packageReferenceArgs.Logger.LogDebug("Project Dependency Graph Read");
+
+            var projectName = dgSpec.Restore.FirstOrDefault();
+            var originalPackageSpec = dgSpec.GetProjectSpec(projectName);
+
+            // Create a copy to avoid modifying the original spec which may be shared.
+            var updatedPackageSpec = originalPackageSpec.Clone();
+            PackageSpecOperations.AddOrUpdateDependency(updatedPackageSpec, packageReferenceArgs.PackageDependency);
+
+            var updatedDgSpec = dgSpec.WithReplacedSpec(updatedPackageSpec).WithoutRestores();
+            updatedDgSpec.AddRestore(updatedPackageSpec.RestoreMetadata.ProjectUniqueName);
+
+            // 2. Run Restore Preview
+            packageReferenceArgs.Logger.LogDebug("Running Restore preview");
+            var restorePreviewResult = await PreviewAddPackageReference(packageReferenceArgs,
```

Comment:
checking Success wont work here, but you should verify GetAllUnresolved() == 0

It is possible the packages were missing or couldn't be downloaded. If that happens they won't show in the compat check since there was nothing to verify. Later it will null ref when they are missing from the graph.